storage/sealing.go

```diff
@@ -0,0 +1,253 @@
+	out := make([]sectorbuilder.PublicPieceInfo, len(t.Pieces))
+	for i, piece := range t.Pieces {
+		out[i] = piece.ppi()
+	}
+	return out
+}
+
+func (t *SectorInfo) deals() []uint64 {
+	out := make([]uint64, len(t.Pieces))
+	for i, piece := range t.Pieces {
+		out[i] = piece.DealID
+	}
+	return out
+}
+
+func (t *SectorInfo) refs() []string {
+	out := make([]string, len(t.Pieces))
+	for i, piece := range t.Pieces {
+		out[i] = piece.Ref
+	}
+	return out
+}
+
+func (t *SectorInfo) existingPieces() []uint64 {
+	out := make([]uint64, len(t.Pieces))
+	for i, piece := range t.Pieces {
+		out[i] = piece.Size
+	}
+	return out
+}
+
+func (t *SectorInfo) rspco() sectorbuilder.RawSealPreCommitOutput {
+	var out sectorbuilder.RawSealPreCommitOutput
+
+	copy(out.CommC[:], t.CommC)
+	copy(out.CommD[:], t.CommD)
+	copy(out.CommR[:], t.CommR)
+	copy(out.CommRLast[:], t.CommRLast)
+
+	return out
+}
+
+func (m *Miner) sectorStateLoop(ctx context.Context) {
+	// TODO: restore state
+
+	go func() {
+		defer log.Warn("quitting deal provider loop")
+		defer close(m.stopped)
+
+		for {
+			select {
+			case sector := <-m.sectorIncoming:
+				m.onSectorIncoming(sector)
+			case update := <-m.sectorUpdated:
+				m.onSectorUpdated(ctx, update)
+			case <-m.stop:
+				return
+			}
+		}
+	}()
+}
+
+func (m *Miner) onSectorIncoming(sector *SectorInfo) {
+	has, err := m.sectors.Has(sector.SectorID)
+	if err != nil {
+		return
+	}
+	if has {
+		log.Warnf("SealPiece called more than once for sector %d", sector.SectorID)
+		return
+	}
+
+	if err := m.sectors.Begin(sector.SectorID, sector); err != nil {
+		// We may have re-sent the proposal
+		log.Errorf("deal tracking failed: %s", err)
+		m.failSector(sector.SectorID, err)
+		return
+	}
+
+	go func() {
+		select {
+		case m.sectorUpdated <- sectorUpdate{
+			newState: api.Packing,
+			id:       sector.SectorID,
+		}:
+		case <-m.stop:
+			log.Warn("failed to send incoming sector update, miner shutting down")
+		}
+	}()
+}
+
+func (m *Miner) onSectorUpdated(ctx context.Context, update sectorUpdate) {
+	log.Infof("Sector %d updated state to %s", update.id, api.SectorStateStr(update.newState))
+	var sector SectorInfo
+	err := m.sectors.Mutate(update.id, func(s *SectorInfo) error {
+		s.State = update.newState
+		if update.mut != nil {
+			update.mut(s)
+		}
+		sector = *s
+		return nil
+	})
+	if update.err != nil {
+		log.Errorf("sector %d failed: %s", update.id, update.err)
+		m.failSector(update.id, update.err)
+		return
+	}
+	if err != nil {
+		m.failSector(update.id, err)
+		return
+	}
+
+	switch update.newState {
+	case api.Packing:
+		m.handle(ctx, sector, m.finishPacking, api.Unsealed)
+	case api.Unsealed:
+		m.handle(ctx, sector, m.sealPreCommit, api.PreCommitting)
+	case api.PreCommitting:
+		m.handle(ctx, sector, m.preCommit, api.PreCommitted)
+	case api.PreCommitted:
+		m.handle(ctx, sector, m.preCommitted, api.SectorNoUpdate)
+	case api.Committing:
+		m.handle(ctx, sector, m.committing, api.Proving)
+	}
```

Comment:
I'd feel more comfortable if there was a default here that logged that something weird was going on